### PR TITLE
APPS-4594 Added validation before adding a new message.

### PR DESCRIPTION
**Pull request by jfernandezsw**
#### Overview

- Ticket: [APPS-4594](https://spryker.atlassian.net/browse/APPS-4594)

### CODE REVIEW

File: src/SprykerSdk/AsyncApi/Code/Builder/AsyncApiBuilder.php [hunk 1]
```diff
@@ -282,9 +282,11 @@ protected function addMessageToChannelType(array $asyncApi, string $messageName,
...
         if (isset($asyncApi['channels'][$channelName][$channelType]['message'])) {
             if (isset($asyncApi['channels'][$channelName][$channelType]['message']['oneOf'])) {
-                $asyncApi['channels'][$channelName][$channelType]['message']['oneOf'][] = [
-                    '$ref' => sprintf('#/components/messages/%s', $messageName),
-                ];
+                if ($this->messageNameExists($messageName, $asyncApi['channels'][$channelName][$channelType]['message']['oneOf']) === false) {
+                    $asyncApi['channels'][$channelName][$channelType]['message']['oneOf'][] = [
+                        '$ref' => sprintf('#/components/messages/%s', $messageName),
+                    ];
+                }
```
Comment (stereomon) [inline src/SprykerSdk/AsyncApi/Code/Builder/AsyncApiBuilder.php:289]: The level of nesting is too much now and even was before. Use early return if possible or any other strategy to have a max nesting of 2 ideally 1.

File: src/SprykerSdk/AsyncApi/Code/Builder/AsyncApiBuilder.php [hunk 2]
```diff
@@ -551,4 +558,21 @@ protected function updateAsyncApi(string $targetFile, array $asyncApi): bool
...
+    /**
+     * @param string $messageName
+     * @param array $messages
+     *
+     * @return bool
+     */
+    protected function messageNameExists($messageName, $messages): bool
+    {
+        foreach ($messages as $message) {
+            if ($message['$ref'] === sprintf('#/components/messages/%s', $messageName)) {
```
Comment (stereomon) [inline src/SprykerSdk/AsyncApi/Code/Builder/AsyncApiBuilder.php:571]: You must use `sprintf('#/components/messages/%s', $messageName)` outside of the loop only once and not per iteration.

File: src/SprykerSdk/AsyncApi/Code/Builder/AsyncApiBuilder.php (hunk 1 above)
Comment (jfernandezsw) [reply src/SprykerSdk/AsyncApi/Code/Builder/AsyncApiBuilder.php]: I changed the logic of the method. It checks if `$asyncApi['channels'][$channelName][$channelType]['message']['oneOf']` does NOT exist and then if the message is in the existing array.

Comment (jfernandezsw) [reply src/SprykerSdk/AsyncApi/Code/Builder/AsyncApiBuilder.php]: I reduced the nesting from 3 levels to 1.
There are no extra checks and the conditionals go from the more abstract to the more concrete.

**Review by stereomon — CHANGES_REQUESTED** [review]

File: src/SprykerSdk/AsyncApi/Code/Builder/AsyncApiBuilder.php [hunk 3]
```diff
@@ -551,4 +559,23 @@ protected function updateAsyncApi(string $targetFile, array $asyncApi): bool
...
         return $this->writeToFile($targetFile, $originAsyncApi);
     }
+
+    /**
+     * @param string $messageName
+     * @param array $messages
+     *
+     * @return bool
+     */
+    protected function messageNameExists($messageName, $messages): bool
```
Comment (stereomon) [inline src/SprykerSdk/AsyncApi/Code/Builder/AsyncApiBuilder.php:569]:
```suggestion
    protected function messageNameExists(string $messageName, array $messages): bool
```

File: tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php [hunk 4]
```diff
@@ -100,4 +101,67 @@ public function testAddMessageReturnsErrorCodeAndPrintsErrorMessagesWhenMessageT
...
+                '--' . SchemaMessageAddConsole::OPTION_MESSAGE_TYPE => 'subscribe',
+                '--' . SchemaMessageAddConsole::OPTION_PROPERTY => ['property:string'],
+                '--' . SchemaMessageAddConsole::OPTION_ASYNC_API_FILE => codecept_data_dir('api/asyncapi/asyncapi.yml'),
+                SchemaMessageAddConsole::ARGUMENT_CHANNEL_NAME => 'channelNameA',
+                SchemaMessageAddConsole::ARGUMENT_OPERATION_ID => 'operationId',
+                SchemaMessageAddConsole::ARGUMENT_MESSAGE_NAME => 'OutgoingMessage',
+            ],
+        );
+
+        // The YAML file is parsed after the message was added
```
Comment (stereomon) [inline tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php:129]: That's obvious and doesn't need to be said again. Only add comments for things that are not obvious.

File: tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php [hunk 5]
```diff
@@ -100,4 +101,67 @@ public function testAddMessageReturnsErrorCodeAndPrintsErrorMessagesWhenMessageT
...
+        $asyncApi = Yaml::parseFile(codecept_data_dir('api/asyncapi/asyncapi.yml'));
+
+        // Assert
+        // The message exists only once in the channel
+        $this->tester->assertMessageExistsOnlyOnceInChannel($asyncApi, 'OutgoingMessage', 'channelNameA', 'subscribe');
```
Comment (stereomon) [inline tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php:134]: The assert method name is good and explains enough no need for this inline comment.

File: tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php [hunk 6]
```diff
@@ -100,4 +101,67 @@ public function testAddMessageReturnsErrorCodeAndPrintsErrorMessagesWhenMessageT
...
+    }
+
+    /**
+     * This test ensures that the array 'oneOf' is not created when trying to add an existing message to a channel that has only one message.
+     *
+     * @return void
+     */
+    public function testAddExistingMessageToChannelWithOneMessageDoesNotCreateOneOfArray(): void
```
Comment (stereomon) [inline tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php:142]: Isn't this exactly the same test as the one before? Both test that a message can only be added once. That we create `oneOf` when we have more than one message is not of interest for this bug.

Or am I getting it wrongly?

**Review by stereomon — CHANGES_REQUESTED** [review]

File: tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php (hunk 6 above)
Comment (jfernandezsw) [reply tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php:140]: I had to add this test to cover this part of `AsyncApiBuilder`:
```
if ($this->messageNameExists($messageName, $messages) === true) {
    return $asyncApi;
}
```
This show that if the array `oneOf` still doesn't exist and the message is not added, the array is not created unnecessarily.

Comment (stereomon) [reply tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php:140]: Then it should stay when its for coverage reasons, the diff between both tests is not obvious. Can you highlight it by a doc block comment maybe?

Comment (jfernandezsw) [reply tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php:140]: Yes, I know that the difference is fuzzy. Is it ok if I put something like this in the docblock?
```
Coverage:
if ($this->messageNameExists($messageName, $messages) === true) {
    return $asyncApi;
}
```

Comment (jfernandezsw) [reply tests/SprykerSdkTest/AsyncApi/Console/SchemaMessageAddConsoleTest.php:145]: I've changed the description of the method and added the test coverage, I think is much better now.

**Pull request merged**